### PR TITLE
Draft - TS Guidelines

### PR DESCRIPTION
**Pull request by Convly**
### What does it do?

Init a TS guide in the documentation to find out about Strapis TS conventions.

### Why is it needed?

Prepare the TS refactoring by adding consistency when writing types & code

### How to test it?

- `cd docs`
- `yarn start`
- Head to `Guides > TypeScript`

### CODE REVIEW

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [inline docs/docs/guides/03-typescript.md:19]:
```suggestion
_Note: In some situations, colocated types can still be exported if needed by other modules_
```
But also, what are those situations?

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [inline docs/docs/guides/03-typescript.md:154]:
```suggestion
In most cases, it is recommended to use regular exports (`export`) instead of the default export (`export default`). Regular exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
```

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [inline docs/docs/guides/03-typescript.md:160]:
```suggestion
## 3. Parameters for Generics
```

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [inline docs/docs/guides/03-typescript.md:187]: I thought we always prefered verbose, we just had an exception to allow single letters if it was self-explanatory.

File: docs/docs/guides/03-typescript.md
Comment (joshuaellis) [inline docs/docs/guides/03-typescript.md:37]: Does this make sense in the context of frontend components? i.e. we may export props from one components to be used to create the props of another, it feels strange that these props wouldn't live with it's component?

File: docs/docs/guides/03-typescript.md
Comment (joshuaellis) [inline docs/docs/guides/03-typescript.md:154]:
```suggestion
In most cases, it is recommended to use named exports (export) instead of the default export (export default). Named exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
```

I believe they're called "named exports"

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [inline docs/docs/guides/03-typescript.md:196]: Not necessary but it would be nice if we could think of an example for how to do this when changing a type definition.

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [inline docs/docs/guides/03-typescript.md:197]: Same here, how would a type definition deprecation look for someone making that change?

File: docs/docs/guides/03-typescript.md
Comment (joshuaellis) [inline docs/docs/guides/03-typescript.md:204]: I think this is potentially too backend orientated to be called "public apis", for example the helper-plugin is a public facing package with numerous public APIs – the props of a component for instance.

**Review by joshuaellis — COMMENTED** [review]
Really nice start ⭐

**Review by innerdvations — COMMENTED** [review]
Nice! Just some minor comments

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:204]: The whole document is definitely too backend oriented. I'll work on that, thanks for the feedback :slightly_smiling_face:

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:154]: That make sense, thanks

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:37]: I think in the context of components it definitely make sense to keep the props colocated with the component

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:19]: This for instance https://github.com/strapi/strapi/pull/17104/files#r1242366691

@joshuaellis I think props falls in this category

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:187]: Don't you consider this to be self-explanatory? (between the name of the type, the extends string & the fact that the type definition is quite straight forward)

File: docs/docs/guides/03-typescript.md
Comment (Convly) [inline docs/docs/guides/03-typescript.md:35]: Now that I think more about this (and taken some time to look at other libraries), I'm wondering whether having regular .ts file for type declaration wouldn't make more sense?
And then keeping .d.ts for module declaration & extension.

My only concern is that it adds some level of complexity + you'll sometimes end up with .d.ts next to .ts which might not be very clear.

What do you think @joshuaellis / @innerdvations / @alexandrebodin ?

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md]: 6913093336838e93042f925aa4fb2c82d0a7db1d

Comment (Convly) [reply docs/docs/guides/03-typescript.md]: 6913093336838e93042f925aa4fb2c82d0a7db1d

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md]: 3b8540879cf2f47138fe1855f8f141666d181f4a

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [reply docs/docs/guides/03-typescript.md:52]: In theory I think you're right. We should be using regular .ts files and then exporting whatever we explicitly want to export in each module in a .d.ts file. Pretty close to what we're already doing but renaming most of our .d.ts -> .ts

However, I don't think it hurts anything to keep them in .d.ts files, especially since we still have so much javascript and jsdoc usage in the codebase, and they tend to be picked up better by editors than regular .ts files. Plus .d.ts enforces that we ONLY include typings in those files and no code can accidentally slip in (although maybe there's a rule for that we could add for the types dir? no idea)

I don't have a strong opinion either way, curious to hear arguments from others though.

Comment (joshuaellis) [reply docs/docs/guides/03-typescript.md:52]: someone did a twitter thread on this – https://twitter.com/mattpocockuk/status/1672918127414333440 https://twitter.com/mattpocockuk/status/1673017324587171842

Comment (Convly) [reply docs/docs/guides/03-typescript.md:52]: Yes I saw this one, but I don't think it fits very well Strapi's context.
I mean, it basically ends up being the pros/cons that we've listed above :confused:

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md]: 36ca09280f3d8d21609b3c66b8c424ed3885d5ca

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md]: I've added it for the second bullet point item 36ca09280f3d8d21609b3c66b8c424ed3885d5ca

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:244]: I've added mention of frontend topics such as the component's props, @joshuaellis don't hesitate to tell me if you see any other topic worth mentioning :eyes:

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [reply docs/docs/guides/03-typescript.md:52]: I'm leaning towards `.ts` at this point for the skipLibCheck, and that it seems to be best practices (after reading through the twitter debate and links there). I don't see any major pros to `.d.ts`

File: docs/docs/guides/03-typescript.md
Comment (innerdvations) [reply docs/docs/guides/03-typescript.md:204]: (I swear I replied to this but now I don't see it)

I mean that I thought we had discussed both options would be ✅ rather than avoiding the first one. If someone wants to write it out, that's just as acceptable, but they don't need to if it's that simple.

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:54]: 🤔  do you have resource explaining the pros&cons of using definition files instead of let's say a `types.ts` file ? for shared types in a folder ?

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:88]: Kind of same question with the ts migration why not just `types/core/uid.ts` ?

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:131]: Will this be enforce with eslint ? If so I would remove them from the guidelines so it only includes things we haven't a lint rule for :)

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:169]: I agree with what's said but that might just be too vague to be in the guideline 🤔 we probably need a specific list for this

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:181]: Could we say "For generic type declaration with a single parameter" 🤔  so we make it clearer we only allow it for pure ts declaration and not inside code.

Same question as above what about eslint ?

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:221]: Reasonable = next major :D. We need to maintain a retro compat interface. 

The only exception is when we need to fix a bug

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [inline docs/docs/guides/03-typescript.md:210]: We can add a callout to verify if a type you change is exported by the package. (I'm wondering how much we could automate that eventually, there might be lib that will check that)

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [reply docs/docs/guides/03-typescript.md:244]: This would apply to the FE with customization of state management most likely.

**Review by alexandrebodin — COMMENTED** [review]
Nice start :) 

- I would add a few guidelines on when to / what to export make public :) 
- We should definitely add the `declare module` topic before v5 release & merging :)
- I think we should add eslint rules & configs that we want in this PR too so we have the "goal" of v5

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:131]: What about keeping this + adding a section about eslint + adding backlinks between both?

Comment (Convly) [reply docs/docs/guides/03-typescript.md:131]: Also, 2 things:
- It's possible to enforce that with an eslint rule
- It's also possible to have a mix that preserves both: `import { type Article, createArticle } from 'smth';`

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:169]: What kind of list are you thinking about?

It's basically: always use export {} except in a few situations like "`For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.`"

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md:181]: > and not inside code.

What about

```ts
function doSomething<T extends string>(value: T): T {
  // ...
}
```

It's code, but I don't think it makes sense to name the generic parameter

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [reply docs/docs/guides/03-typescript.md]: Great news. If we ever change the rule the guidelines won't be up to date. It's really hard to maintain those kind of things that less is better IMO. I would instead add a section that link to the strapi commong eslint config and the local config file to know about all the others rules

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [reply docs/docs/guides/03-typescript.md]: Yes these are just examples :) we need clear rules with clear exceptions :)

File: docs/docs/guides/03-typescript.md
Comment (alexandrebodin) [reply docs/docs/guides/03-typescript.md]: Well yup +1

File: docs/docs/guides/03-typescript.md
Comment (Convly) [reply docs/docs/guides/03-typescript.md]: I've tried to make it clearer, not sure if this is what you meant. Waiting for your feedback

Comment (Convly) [reply docs/docs/guides/03-typescript.md]: [eff4e40](https://github.com/strapi/strapi/pull/17104/commits/eff4e403216736b51cf3ef9bc4fd281446adfe5e)

Comment (alexandrebodin) [reply docs/docs/guides/03-typescript.md]: Should be fine for now

**Review by alexandrebodin — APPROVED** [review]

**Comment by alexandrebodin** [comment]
Closing for now ;) feel free to resubmit when you are back

**Pull request closed without merging**
